org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/DocumentSymbolService.java

```diff
@@ -99,12 +113,43 @@ public List<? extends Location> getDefinitions(Document document, XtextResource
 		return getDefinitions(resource, offset, resourceAccess, cancelIndicator);
 	}
 
+	/**
+	 * @since 2.27
+	 */
+	protected EObject getElementWithIdentifierAt(XtextResource xtextResource, int offset) {
+		if (offset >= 0) {
+			if (xtextResource != null) {
+				IParseResult parseResult = xtextResource.getParseResult();
+				if (parseResult != null) {
+					ICompositeNode rootNode = parseResult.getRootNode();
+					if (rootNode != null) {
+						ILeafNode leaf = NodeModelUtils.findLeafNodeAtOffset(rootNode, offset);
+						if (leaf != null && isIdentifier(leaf)) {
+							return eObjectAtOffsetHelper.resolveElementAt(xtextResource, offset);
+						}
+					}
+				}
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * @since 2.27
+	 */
+	protected boolean isIdentifier(ILeafNode leafNode) {
+		return (leafNode.getGrammarElement() instanceof RuleCall
```

Comment:
will this work with name="keyword"?

Reply:
I am not sure what you have in mind. It works for the test case 4, where the cursor is at the beginning of "type", which is a keyword.

Reply:
if i have in grammar something like

Variable: "var" name=XXX;
or 
VariableInSpecialContext returns Variable: "var" name="thereyougo";
will it work?

Reply:
I am not sure about what grammar element would that produce, but I think it would because 
a) it works in the test I added for the type
TypeDeclaration:
    'type' name=ID ...
;
and I think that is the same structure as the variable you sketched.
b) if it would not, we would have the same problem for the existing rename service (I guess we do not because no one complained so far)

Reply:
I have found some time to move the code to where I would refactor it (EObjectAtOffsetHelper and a new class called NodeHelper). I still need to work on the other points.
Regards

Reply:
Hi @szarnekow and @cdietrich,

I have now a version that will work as well with
`VariableInSpecialContext` returns Variable: "var" `name="thereyougo";`
but will still not let you rename it.

I am afraid I did not get your question about the offset being part of a data type rule, could you please explain that point?

Thanks

Reply:
@szarnekow , I have checked by modifying the TestLanguage.xtext to include a data type rule and tested the change. The code I have now in the PR would work for data types rules.
I think I have addressed now all the points. 
Regards